[Experiment] C++ interface to create a lowered backend module.

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51232 [Experiment] C++ interface to create a lowered backend module.**

Differential Revision: [D26139281](https://our.internmc.facebook.com/intern/diff/D26139281)

iseeyuan committed:
[Experiment] C++ interface to create a lowered backend module.

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -48,6 +52,221 @@ c10::FunctionSchema getExecuteSchema() {
+  // Generate LoweredModule.
+  Module loweredModule(
+      "torch.jit." + backend_name + "LoweredModule",
+      get_python_cu(),
```
We need to get rid of this. I think this is the only remaining dependency on Python. Unfortunately, this returns a pointer to an object that is defined in Python. For the purposes of the experiment, I think we can add another parameter to the function of type `std::shared_ptr<CompilationUnit>`. In Python, the caller should pass `get_python_cu()`. In C++, the caller will have to use a compilation unit created by loading a module (see source code for `torch.jit.save`) or make their own.

iseeyuan committed:
Update on "[Experiment] C++ interface to create a lowered backend module."




[ghstack-poisoned]

iseeyuan committed:
Update on "[Experiment] C++ interface to create a lowered backend module."




[ghstack-poisoned]

iseeyuan committed:
Update on "[Experiment] C++ interface to create a lowered backend module."




[ghstack-poisoned]

iseeyuan commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -337,6 +338,29 @@ class TorchBindLiteInterpreterTestStruct
+
+  auto minput = 5 * torch::ones({});
+  auto ref = m.run_method("forward", minput).toTensor().item<float>();
+  auto res = lowered_m.run_method("forward", minput);
```
I got error from this line:
C++ exception with description "The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
RuntimeError: Expected 1 elements in a list but found 2
" thrown in the test body.
@SplitInfinity , anything I did wrong?

SplitInfinity replied:
I think this exception is caused by the list unpack operator generated by this line of code inside `codegen_backend_module`:
```
$unpack, = self.__backend.execute(self.__handles["$method"], typed_inputs)
```

I noticed that you used the `test_backend` when calling `codegen_backend_module`. The test backend that we use to test this feature is very basic and is hardcoded to return two outputs for `forward`. You can see the implementation in `test/cpp/jit/test_backend.cpp` (relative to root). I would recommend implementing your own backend that is more intelligent or implements exactly the ops you are testing.

iseeyuan replied:
Aha! That makes sense. I'll implement the backend.

iseeyuan committed:
Update on "[Experiment] C++ interface to create a lowered backend module."




[ghstack-poisoned]

iseeyuan committed:
Update on "[Experiment] C++ interface to create a lowered backend module."


Differential Revision: [D26139281](https://our.internmc.facebook.com/intern/diff/D26139281)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -341,14 +348,15 @@ c10::IValue BytecodeDeserializer::readArchive(
 
-  static const c10::QualifiedName torchPrefix = "__torch__";
+  static const c10::QualifiedName torchPrefix__ = "__torch__";
+  static const c10::QualifiedName torchPrefix = "torch";
```
@SplitInfinity and @suo , it looks hacky here to just work around runtime error. The error happens when readArchive("data.pkl"), because the type parser in lite interpreter cannot resolve type with token "torch". data.pkl looks like:
```
torch.jit.mul_test_backendLoweredModule()(state=
 (torch.jit._pickle.restore_type_tag({'forward': torch.jit._pickle.restore_type_tag({'': ''}, 'Dict[str, Any]')},
    'Dict[str, Any]'),
  __torch__.___torch_mangle_0.M()(state={})))
```
Looks like it's used to describe the module itself, but it's not used at lite interpreter runtime. Do you have any better idea to work around?

SplitInfinity replied:
Unfortunately I don't know *anything* about the details of pickling/unpickling 🙉 . I can dig some more next week if @suo doesn't know.

iseeyuan committed:
Update on "[Experiment] C++ interface to create a lowered backend module."


Differential Revision: [D26139281](https://our.internmc.facebook.com/intern/diff/D26139281)

[ghstack-poisoned]

iseeyuan committed:
Update on "[Experiment] C++ interface to create a lowered backend module."


Differential Revision: [D26139281](https://our.internmc.facebook.com/intern/diff/D26139281)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/backends/backend_detail.cpp:
```diff
@@ -48,6 +51,221 @@ c10::FunctionSchema getExecuteSchema() {
+                self.__method_compile_spec = state[0]
+                self.__processed_module = state[1]
+                self.__create_backend()
+                self.__handles = self.__backend.compile(self.__processed_module, self.__method_compile_spec)
```
So at the moment we will do compilation at load time? But we do need to be able to this AOT right? And if so I am wondering how are we going to store compiled methods, or whatever else the processed_module needs to store, that does not have native serialization support in pytorch. Do we cast it to std::vector<uint8_t> or something like that?

iseeyuan replied:
Sorry for some reason I missed this comment. Probably it's an experimental PR and I was focused on the "formal" ones. You might already got your answers from #51232

kimishpatel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -337,6 +340,121 @@ class TorchBindLiteInterpreterTestStruct
+  compile_spec.insert("forward", fake_dict);
+  auto any_dict_ty = DictType::create(StringType::get(), AnyType::get());
+  // lowered module
+  auto lm = torch::jit::detail::codegen_backend_module(
```
Shouldnt we test via `_jit_to_backend("mul_test_backend", m, compile_spec)`?

kimishpatel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -827,6 +946,60 @@ static auto reg =
+    at::Tensor output = value.toTensor();
+    output = output.clone();
+
+    if (handle.toStringRef() == "forward") {
```
So this is somewhat decoupled from what the module under test defines in ToBackend. I would suggest to do this if possible:
1. In preprocess just "compile" forward method of the module by lowering it to bytecode. Using `getFunctionTuple` defined in export_module.cpp. This way we get the bytecode corresponding to that method and that can be serialized as part of the module?. Output of the `getFunctionTuple` can be stored as a new attribute of the returned module such as `compiled_methods`.
2. Then in `compile` part of the backend, extract `compiled_methods` attribute and then construct `Code` out of it the way we do in lite interpreter.

This probably is little too long of a path but thought it might showcase an example of how to serialize compiled methods.
